assets/shared/helpers/player.js

```diff
@@ -0,0 +1,539 @@
+			};
+		},
+	},
+};
+
+/**
+ * A class that abstracts the use of the player APIs: Video, VideoPress, YouTube, and Vimeo.
+ */
+class Player {
+	/**
+	 * Player constructor.
+	 *
+	 * @param {HTMLVideoElement|HTMLIFrameElement} element The player element.
+	 */
+	constructor( element ) {
+		this.playerPromise = null;
+		this.type = null;
+		this.element = element;
+
+		try {
+			this.setType( element );
+		} catch ( e ) {
+			// eslint-disable-next-line no-console -- We want to expose the element with problem.
+			console.error( e, element );
+		}
+	}
+
+	/**
+	 * Set the player type.
+	 *
+	 * @throws Will throw an error if the video type is not found.
+	 */
+	setType() {
+		if ( this.element instanceof window.HTMLVideoElement ) {
+			this.type = VIDEO_TYPE;
+		} else if ( this.element instanceof window.HTMLIFrameElement ) {
+			this.type = Object.entries( players ).find(
+				( [ , p ] ) =>
+					p.embedPattern && this.element.src.match( p.embedPattern )
+			)?.[ 0 ];
+		}
+
+		if ( ! this.type ) {
+			throw new Error( 'Video type not found' );
+		}
+	}
+
+	/**
+	 * Get the video player.
+	 *
+	 * @return {Promise<Object|HTMLVideoElement|HTMLIFrameElement>} The video player through a promise.
+	 */
+	getPlayer() {
+		if ( ! this.playerPromise ) {
+			this.playerPromise =
+				players[ this.type ]?.initializePlayer( this.element ) ||
+				// A promise that never resolves if it doesn't exist.
+				Promise.reject( new Error( 'Failed getting the player' ) );
+		}
+
+		return this.playerPromise;
+	}
+
+	/**
+	 * Get the video duration.
+	 *
+	 * @return {Promise<number>} The duration of the video in seconds through a promise.
+	 */
+	getDuration() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].getDuration( player )
+		);
+	}
+
+	/**
+	 * Set the video to a current time.
+	 *
+	 * @param {number} seconds The video time in seconds to set.
+	 *
+	 * @return {Promise} A promise that resolves if the video was set to a current time successfully.
+	 */
+	setCurrentTime( seconds ) {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].setCurrentTime( player, seconds )
+		);
+	}
+
+	/**
+	 * Play the video.
+	 *
+	 * @return {Promise} A promise that resolves if the video was played successfully.
+	 */
+	play() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].play( player )
+		);
+	}
+
+	/**
+	 * Pause the video.
+	 *
+	 * @return {Promise} A promise that resolves if the video was paused successfully.
+	 */
+	pause() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].pause( player )
+		);
+	}
+
+	/**
+	 * Add an event listener to the player.
+	 *
+	 * @param {string}   eventName Event name (supported: `timeupdate`).
+	 * @param {Function} callback  Listener callback.
+	 *
+	 * @throws Will throw an error if the event is not supported.
+	 *
+	 * @return {Promise<Function>} The function to unsubscribe the event through a promise.
```

Comment:
I started implementing it as `on` and `off` methods, but I changed the approach in order to simplify the code.

Since we need to enhance the `callback` to every player, we'd need to memorize the new `callback`, so we could remove the event listener later. And it would make the code much more complicated, so I just returned the function to unsubscribe it from the `on`.

Reply:
Probably we'll want others in the future, but since for now its just this one, I think it makes sense: https://github.com/Automattic/sensei/pull/5364#discussion_r929259028